### DISCUSSION TITLE
Enforcing usage of catalog dependencies

### DISCUSSION THREAD

**Discussion by rhinodavid** [pnpm/pnpm#8567] [Q&A] [answered]
I was excited to see the release of [workspace catalogs](https://pnpm.io/catalogs).

My team has been using [syncpack](https://github.com/JamieMason/syncpack) to accomplish syncing dependency versions throughout our monorepo, and having that functionality in the package manager seems appropriate.

There's one feature I think is missing, however. We run `syncpack list-mismatches` in CI to make sure no workspace is using a version of a dependency other than the one specified in the syncpack configuration. Similarly, I'd like to have a way to make sure that if a dependency is specified in the catalog, no workspace in the repo uses a version other than `catalog:` (or a named catalog).

Firstly, I just want to make sure I'm not missing functionality that might do that now.

If not, I'm wondering if anyone else would be interested in either an external tool to do this, or for a command to enforce catalog usage to be added to the `pnpm` CLI.

**Comment by theoludwig** [comment]
That would be a cool addition, to enforce usage of workspace catalogs in a monorepo, and forbid usage of dependencies versions directly.

**Comment by JamieMason** [comment]
Hello 👋 From a read of the docs, it looks like syncpack could possibly support this if you'd like to write up an issue @rhinodavid. It already reads pnpm-workspace.yaml so it could bring the catalogs data in and try and make use of it.

**Comment by zkochan** [comment]
We can probably support that but first need to add catalog support to the `pnpm add` command and `pnpm update`.

**Reply by thdk** [reply]
I'm new to pnpm and also previously used syncpack in CI to ensure consistent use of all dependency versions. I would like to follow up on this to see what still remains open.

I believe both `pnpm add` and `pnpm update` now support the `catalog:` protocol.

- Merged PR for `pnpm add` command: https://github.com/pnpm/pnpm/pull/8870.
- Merged PR for `pnpm update` command: https://github.com/pnpm/pnpm/pull/9517

Also semi related is that there is currently a[ PR in progress at `npm-check-updates` ](https://github.com/raineorshine/npm-check-updates/pull/1533)to also add catalog support for pnpm users.

**Reply by zkochan** [reply]
This was implemented. You can set the [catalogMode: strict](https://pnpm.io/catalogs#catalogmode) setting

**Comment by zkochan** [comment, marked as answer]
This was implemented. You can set the [catalogMode: strict](https://pnpm.io/catalogs#catalogmode) setting

**Reply by ryuuji3** [reply]
If someone hand-edits `package.json` this `catalogMode` won't apply right? Should it? IMO yes.

**Comment by faithmckinely-alt** [comment]
It should I guess

**Comment by JamieMason** [comment]
Hey @rhinodavid,
[syncpack 15.0.0](https://github.com/JamieMason/syncpack/releases/tag/15.0.0) has new functionality to help migrate to, manage, and update dependencies in your pnpm catalogs. 

More information is in the [release notes](https://github.com/JamieMason/syncpack/releases/tag/15.0.0) I have linked, hopefully this is helpful, thanks.
